lean4-infoview-api/src/infoviewApi.ts

```diff
@@ -128,3 +129,9 @@ export interface InfoviewApi {
    */
   requestedAction(action: InfoviewAction): Promise<void>;
 }
+
+export interface TestApi {
```

Comment:
Could you please explain what `TestApi` is for? Is this functionality that the infoview itself provides for test purposes? If so, how could `isInfoViewOpen` possibly return `false`? On the other hand if the `TestApi` isn't so much methods on the infoview as much as it is something for extension<->test suite communication, then it should live in the `vscode-lean4` package.

Copying the infoview HTML to clipboard seems like quite wierd functionality, but if it's needed for something in testing, fine by me. `toggleAllMessages` seems to duplicate `requestedAction('toggleAllMessages')` so probably isn't needed?

Reply:
Judging from the `messages.tsx` diff, only `copyHtmlToClipboard` is functionality that actually needs to be added here, so I would put it directly in `InfoviewApi` and put a comment explaining what kind of test it's needed for.

Reply:
Sure, it is an Api that allows the "test" package to be strongly typed on an interface.  We could program it all using `any` but that's not very nice.  So we have this handy infoview api package which the test can take a dependency on.  The other approach is to have the test depend on vscode-lean4 entirely then we could access all the types defined in there and not need to "funnel" everything through an explicit API, but then again I kinda like being explicit about what the test code needs.  

The Copy HTML is the most generic thing I could think of that allows the tests do look for any thing they want to in the info view so I don't have to keep adding more and more TestApi methods.  toggleAllMessages  is indeed a wrapper on requestedAction('toggleAllMessages'), again in the spirit of being strongly typed and explicit about what the tests are depending on.  Conversely if the test depended on vscode-lean4 it could depend on `InfoProvider` and call that guy's toggleAllMessages method...  So if you want me to drop the TestApi idea and just expose all of vscode-lean4 to the test package then let me know and I can do that.  It probably means eventually the object returned from `activate` will return also return the InfoProvider, DocViewProvider, AbbreviationFeature, LeanClientProvider, etc, and then the tests can go to town with those objects...

Reply:
Having an explicit `TestApi` sounds like a good idea, I am just not sure it should live in `lean4-infoview` since that package is meant to be independent of VSCode and extensions for other editors may want to test differently. The crux of the issue seems to be that the `test` package does not depend on `vscode-lean4`, otherwise `TestApi` could live there. This is indeed like the setup of `helloworld-test-sample`. On the other hand the Yeoman extension scaffold  (referenced [here](https://code.visualstudio.com/api/working-with-extensions/testing-extension#overview)) places integration tests directly in the extension package and does not bother to create a separate test package. Maybe just doing that would be simpler? The Haskell extension [does it](https://github.com/haskell/vscode-haskell/tree/79456aaaa98b98f9e7b9f690bb18ab2b2838b6e2/test). In particular, I can't see anyone wanting to depend on `vscode-lean4` but not the tests for some reason.

Reply:
But we could rename the `lean4-infoview-api` and call it `vscode-lean4-api` then it wouldn't look so weird...

Reply:
Well, what do you think about moving `test` to the `vscode-lean4` package like Haskell's VSC extension? Note that `lean4-infoview-api` is also intended not to depend on VSCode.

Reply:
I'm not against it, @gabriel, is this what you recommend also?

Reply:
> Having an explicit TestApi sounds like a good idea, I am just not sure it should live in lean4-infoview since that package is meant to be independent of VSCode and extensions for other editors may want to test differently.

We can also rename it `VscodeTestApi`.  I agree that it doesn't really fit in that package, but there's no technical downside to putting it there either.

> In particular, I can't see anyone wanting to depend on vscode-lean4 but not the tests for some reason.

Nobody is *depending* on the `vscode-lean4` package.  The issue here is whether the tests (and their dependencies) are included in the packaged vscode extension.  But I think webpack does tree shaking already (that's why you don't see the test code in the generated extension.js file).

> `import { InfoProvider } from '../../../../vscode-lean4/src/infoview';`

This works just fine once you remove the extra `tsconfig.json` file in the `test` directory.

If we can have the tests in the `test` subdirectory of the `vscode-lean4` package, then that's the most straightforward solution.  I'd preferably avoid adding new packages if possible; the monorepo setup is already complicated enough as it is.

Reply:
> The issue here is whether the tests (and their dependencies) are included in the packaged vscode extension. 

They wouldn't be -- this is configurable in `.vscodeignore`.

> If we can have the tests in the test subdirectory of the vscode-lean4 package, then that's the most straightforward solution. I'd preferably avoid adding new packages if possible

I agree and would be happy with this also -- as long as tests aren't in a separate package, I am agnostic as to whether they reside under `vscode-lean4/src/test`, or `vscode-lean4/test`, or `test/`. Maybe either of the first two is still preferable since ultimately only `vscode-lean4` is being integration-tested.

Reply:
So I'm still messing with this but the "webpack" job is getting in the way.  If I'm going to merge the test in with vscode_lean4, then I need to figure out how to break the compile into 2 stages, a "webpack" job for "src/**/*" and a normal "npm run compile" for "test/**/*" with a different "test/out" folder.  I'm messing with https://stackoverflow.com/questions/37579969/how-to-use-multiple-tsconfig-files-in-vs-code but no luck yet.  Do either of you have experience with this sort of thing?  If I can't get it to work, I will have to use the separate test project.

Reply:
IF you want to help, perhaps you can experiment with https://github.com/lovettchris/vscode-lean4/tree/clovett/merge_unittest

Reply:
Oh, hey, I figured it out, the tests are working, they are still in a separate project but it does not need a TestApi.  on reading https://www.typescriptlang.org/docs/handbook/project-references.html more closely I discovered a new build option `--build` is needed on the tsc build jobs instead of `tsc -p ./` it is now `tsc --build` and `tsc --build --watch`, so now the command line tests pass and the vs debugging of tests also works.

Reply:
But rats, the same `tsc --build` that works on my machine is not working in the CI build...

Reply:
If you put the tests in a separate package, then you can just list the extension as a dependency in test/package.json.  There's no need to bother with project references.

Reply:
Ok, Wojciech helped me out, and we have something working without a separate test package.